controllers/controlplaneupgrade_controller.go

```diff
@@ -239,42 +242,84 @@ func (r *ControlPlaneUpgradeReconciler) updateMachine(ctx context.Context, log l
 	if err != nil {
 		return err
 	}
+
+	log = log.WithValues("Machine", machine.Name)
 	machinePatchHelper, err := patch.NewHelper(machine, r.client)
 	if err != nil {
 		return err
 	}
-	log.Info("Updating K8s version and kubeadmClusterConfiguration annotation in machine", "Machine", machine.Name)
-	// Update the machine k8s version
-	machine.Spec.Version = &nodeUpgrade.Spec.KubernetesVersion
 
-	// Update the machine kubeadmClusterConfiguration annotation
-	kcc, err := getKubeadmClusterConfig(cpUpgrade)
+	log.Info("Updating K8s version and kubeadmClusterConfiguration annotation in machine")
+	kcpSpec, err := decodeAndUnmarshalKcpSpecData(cpUpgrade.Spec.ControlPlaneSpecData)
 	if err != nil {
 		return err
 	}
-	annotations.AddAnnotations(machine, map[string]string{kubeadmClusterConfigurationAnnotation: kcc})
+
+	// Update the machine kubeadmClusterConfiguration annotation
+	kcc, err := json.Marshal(kcpSpec.KubeadmConfigSpec.ClusterConfiguration)
+	if err != nil {
+		return fmt.Errorf("marshaling KCP cluster configuration: %v", err)
+	}
+
+	// Update the machine k8s version and update the KubeadmClusterConfiguration annotation
+	machine.Spec.Version = &nodeUpgrade.Spec.KubernetesVersion
+	annotations.AddAnnotations(machine, map[string]string{kubeadmClusterConfigurationAnnotation: string(kcc)})
 
 	if err := machinePatchHelper.Patch(ctx, machine); err != nil {
 		return fmt.Errorf("updating spec for machine %s: %v", machine.Name, err)
 	}
 
+	if err := r.updateKubeadmConfig(ctx, log, kcpSpec, machine); err != nil {
+		return fmt.Errorf("updating kubeadm config: %v", err)
+	}
+
 	return nil
 }
 
-func getKubeadmClusterConfig(cpUpgrade *anywherev1.ControlPlaneUpgrade) (string, error) {
-	kcpSpec := &controlplanev1.KubeadmControlPlaneSpec{}
-	decodedKcpSpec, err := base64.StdEncoding.DecodeString(cpUpgrade.Spec.ControlPlaneSpecData)
+func (r *ControlPlaneUpgradeReconciler) updateKubeadmConfig(ctx context.Context, log logr.Logger, kcpSpec *controlplanev1.KubeadmControlPlaneSpec, machine *clusterv1.Machine) error {
+	bootstrapRef := machine.Spec.Bootstrap.ConfigRef
+	if bootstrapRef == nil {
+		return fmt.Errorf("bootstrap config for machine %s is nil", machine.Name)
+	}
+
+	kc := &bootstrapv1.KubeadmConfig{}
+	if err := r.client.Get(ctx, types.NamespacedName{Name: bootstrapRef.Name, Namespace: machine.Namespace}, kc); err != nil {
+		return fmt.Errorf("retrieving bootstrap config for machine %s: %v", machine.Name, err)
+	}
+
+	patchHelper, err := patch.NewHelper(kc, r.client)
 	if err != nil {
-		return "", fmt.Errorf("decoding cpUpgrade.Spec.ControlPlaneSpec: %v", err)
+		return fmt.Errorf("initializing patch helper for KubeadmConfig %s: %v", kc.Name, err)
 	}
-	if err := json.Unmarshal(decodedKcpSpec, kcpSpec); err != nil {
-		return "", fmt.Errorf("unmarshaling cpUpgrade.Spec.ControlPlaneSpec: %v", err)
+
+	kcSpec := kcpSpec.KubeadmConfigSpec.DeepCopy()
+	if kc.Spec.InitConfiguration == nil {
```

Comment:
no, it's because `kcp.Spec. KubeadmConfigSpec` config has both the init and join configs but the `KubeadmConfig` only has either init or join depending whether it's the first control plane node or not.
So, we should only update the `KubeadmConfig`'s init or join config section based on whichever one was set